examples/defaulter-gen/generators/defaulter.go

```diff
@@ -0,0 +1,710 @@
+				for _, existing := range v.additional {
+					if existing.Name == f.Name {
+						exists = true
+						break
+					}
+				}
+				if exists {
+					continue
+				}
+			}
+			v.additional = append(v.additional, f)
+			manualMap[key] = v
+			glog.V(6).Infof("found additional defaulter function for %s from %s", key.Name, f.Name)
+		}
+		buffer.Reset()
+		sw.Do("$.inType|objectdefaultfn$", args)
+		if f.Name.Name == buffer.String() {
+			key := inType.Elem
+			// We might scan the same package twice, and that's OK.
+			v, ok := manualMap[key]
+			if ok && v.base != nil && v.base.Name.Package != pkg.Path {
+				panic(fmt.Sprintf("duplicate static defaulter defined: %#v", key))
+			}
+			v.object = f
+			manualMap[key] = v
+			glog.V(6).Infof("found object defaulter function for %s from %s", key.Name, f.Name)
+		}
+		buffer.Reset()
+	}
+}
+
+func Packages(context *generator.Context, arguments *args.GeneratorArgs) generator.Packages {
+	boilerplate, err := arguments.LoadGoBoilerplate()
+	if err != nil {
+		glog.Fatalf("Failed loading boilerplate: %v", err)
+	}
+
+	packages := generator.Packages{}
+	header := append([]byte(fmt.Sprintf("// +build !%s\n\n", arguments.GeneratedBuildTag)), boilerplate...)
+	header = append(header, []byte(
+		`
+// This file was autogenerated by defaulter-gen. Do not edit it manually!
+
+`)...)
+
+	// Accumulate pre-existing default functions.
+	// TODO: This is too ad-hoc.  We need a better way.
+	existingDefaulters := defaulterFuncMap{}
+
+	buffer := &bytes.Buffer{}
+	sw := generator.NewSnippetWriter(buffer, context, "$", "$")
+
+	// We are generating defaults only for packages that are explicitly
+	// passed as InputDir.
+	for _, i := range context.Inputs {
+		glog.V(5).Infof("considering pkg %q", i)
+		pkg := context.Universe[i]
+		if pkg == nil {
+			// If the input had no Go files, for example.
+			continue
+		}
+
+		// Add defaulting functions.
+		getManualDefaultingFunctions(context, pkg, existingDefaulters)
+
+		var peerPkgs []string
+		if customArgs, ok := arguments.CustomArgs.(*CustomArgs); ok {
+			if len(customArgs.ExtraPeerDirs) > 0 {
+				peerPkgs = append(peerPkgs, customArgs.ExtraPeerDirs...)
+			}
+		}
+		// Make sure our peer-packages are added and fully parsed.
+		for _, pp := range peerPkgs {
+			context.AddDir(pp)
+			getManualDefaultingFunctions(context, context.Universe[pp], existingDefaulters)
+		}
+
+		typesWith := extractTag(pkg.Comments)
+		shouldCreateObjectDefaulterFn := func(t *types.Type) bool {
+			if defaults, ok := existingDefaulters[t]; ok && defaults.object != nil {
+				// A default generator is defined
+				glog.V(5).Infof("  an object defaulter already exists as %s", defaults.base.Name)
+				return false
+			}
+			// opt-out
+			if checkTag(t.SecondClosestCommentLines, "false") {
+				return false
+			}
+			// opt-in
+			if checkTag(t.SecondClosestCommentLines, "true") {
+				return true
+			}
+			// type contains a field with a type name matching the pkg comment tag
```

Comment:
Conversion assumes all objects need conversion in a package except those that opt out (reasonable - we match everything to everything).  OpenAPI is similar.  Defaulting, on the other hand, is used almost exclusively on top level objects, and our defaulters today do not "cover" (to properly default you have to invoke all nested defaulters at the right spot).

This code generates top level covering defaulters for everything that needs them.  So PodSpec doesn't really need a covering defaulter - but Pod does.  So instead of the blanket package inclusion + selective exclusion, the defaulter supports selective inclusion/exclusion per type, and then blanket selectors based on whether you have a field.  In Kube, conveniently, every type we want a top level defaulter for has TypeMeta.  Rather than hardcode type meta, I allow it to be defined as a field.  So all Lists + Objects + quasi-objects (RangeAllocations) etc get object defaulters.